include/deal.II/base/thread_local_storage.h

```diff
@@ -158,29 +153,25 @@ namespace Threads
     void
     clear();
 
+  private:
     /**
-     * Return a reference to the internal Threading Building Blocks
-     * implementation. This function is really only useful if deal.II has been
-     * configured with multithreading and has no useful purpose otherwise.
+     * The data element we store.
      */
-#  ifdef DEAL_II_WITH_THREADS
-    tbb::enumerable_thread_specific<T> &
-#  else
-    T &
-#  endif
-    get_implementation();
+    std::map<std::thread::id, T> data;
 
-  private:
-#  ifdef DEAL_II_WITH_THREADS
     /**
-     * The data element we store. If we support threads, then this object will
-     * be of a type that provides a separate object for each thread.
-     * Otherwise, it is simply a single object of type T.
+     * A mutex to guard insertion into the data object.
      */
-    tbb::enumerable_thread_specific<T> data;
+#  ifdef DEAL_II_HAVE_CXX17
+    std::shared_mutex insertion_mutex;
 #  else
-    T data;
+    std::shared_timed_mutex insertion_mutex;
```

Comment:
Nice idea. I had not known that C++14 had introduced these read/write locks :-) 

I think it would be useful to mention in the documentation why we use this kind of mutex here.

Reply:
Funky new C++14 feature, isn't it? :-)